httpbp/middlewares.go

```diff
@@ -266,3 +274,103 @@ func SupportedMethods(method string, additional ...string) Middleware {
 		}
 	}
 }
+
+// statusCodeRecorder is used by RecordStatusCode to record the code passed to
+// a call to WriteHeader.
+type statusCodeRecorder struct {
+	http.ResponseWriter
+
+	code int
+}
+
+func (r *statusCodeRecorder) WriteHeader(code int) {
+	r.ResponseWriter.WriteHeader(code)
+	if r.code == 0 {
+		r.code = code
+	}
+}
+
+// getStatusCodeFamily takes an http status code and returns it as an "Nxx"
+// string. Returns "nan" if code > 999 and "-nan" if code < 100.
+func getStatusCodeFamily(code int) string {
+	family := code / 100
+	switch family {
+	case 1:
+		return "1xx"
+	case 2:
+		return "2xx"
+	case 3:
+		return "3xx"
+	case 4:
+		return "4xx"
+	case 5:
+		return "5xx"
+	case 6:
+		return "6xx"
+	case 7:
+		return "7xx"
+	case 8:
+		return "8xx"
+	case 9:
+		return "9xx"
+	default:
+		if family < 1 {
+			return "-nan"
+		}
+		return "nan"
+	}
+}
+
+// CounterGenerator is used by RecordStatusCode to create counters for recording
+// http response codes set by the server.
+type CounterGenerator interface {
+	Counter(name string) metrics.Counter
+}
+
+// RecordStatusCode extracts the status code set on the request in the following
+// order:
+//	1. Check if WriteHeader was called on the ResponseWriter and use that code
+//	if it was.
+//	2. If an error was returned, check if it is an HTTPError. If it is, use the
+//	code from the error, otherwise assume 500.
+//	3. Assume 200.
+//
+// If it sees an invalid status code (<100 or >999), it will record the status
+// as "-nan" for codes <100 and "nan" for codes >999. Note that a code that is
+// <100 or >999 is unlikely to appear here and will cause a  panic if passed to
+// WriteHeader.
+//
+// RecordStatusCode should generally not be used directly, instead use the
+// NewBaseplateServer function which will automatically include RecordStatusCode
+// as one of the Middlewares to wrap your handlers in.
+func RecordStatusCode(counters CounterGenerator) Middleware {
+	return func(name string, next HandlerFunc) HandlerFunc {
+		counter := counters.Counter("baseplate.http." + name + ".response")
+		return func(ctx context.Context, w http.ResponseWriter, r *http.Request) (err error) {
+			wrapped := &statusCodeRecorder{ResponseWriter: w}
+			defer func() {
+				var code int
+				if wrapped.code != 0 {
+					// WriteHeader was called explicitly, use that
+					code = wrapped.code
+				} else if err != nil {
+					// something went wrong, check if err is an HTTPErr where we can extract
+					// the code, otherwise assume InternalServerError
+					var httpErr HTTPError
+					if errors.As(err, &httpErr) {
+						code = httpErr.Response().Code
+					} else {
+						code = http.StatusInternalServerError
+					}
+				} else {
+					// if there's no error returned and no call to WriteHeader, assume OK
```

Comment:
slight nit on the "assume" part: this is actually clearly documented that if `WriteHeader` was never called it will be 200, so we can just link the doc here in the comment: https://pkg.go.dev/net/http#ResponseWriter.WriteHeader